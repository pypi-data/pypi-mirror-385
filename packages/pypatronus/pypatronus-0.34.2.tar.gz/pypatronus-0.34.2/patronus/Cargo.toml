[package]
name = "patronus"
version.workspace = true
description = "Hardware bug-finding toolkit."
homepage = "https://kevinlaeufer.com"
keywords = ["RTL", "btor", "model-checking", "SMT", "bit-vector"]
edition.workspace = true
authors.workspace = true
repository.workspace = true
readme = "Readme.md"
license.workspace = true
rust-version.workspace = true

[dependencies]
indexmap = "2.7.0"
codespan-reporting = "0.11.1"
lazy_static = "1.4.0"
smallvec = { version = "1.x", features = ["union"] }
rand = { version = "0.9", features = ["small_rng"] }
boolean_expression.workspace = true
regex.workspace = true
baa.workspace = true
rustc-hash.workspace = true
thiserror.workspace = true
fst-writer = "0.3.1"

[dev-dependencies]
insta = { version = "1.x", features = ["yaml"] }

