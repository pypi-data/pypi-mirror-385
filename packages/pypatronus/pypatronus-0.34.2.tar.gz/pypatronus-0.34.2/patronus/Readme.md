# Patronus


[![Crates.io Version](https://img.shields.io/crates/v/patronus)](https://crates.io/crates/patronus)
[![docs.rs](https://img.shields.io/docsrs/patronus)](https://docs.rs/patronus)
[![GitHub License](https://img.shields.io/github/license/ekiwi/patronus)](LICENSE)



### TODO

Some things we will hopefully get to one day.

- simulator
  - JIT based implementation
  - better debugging, add option to print expressions with trace
  - waveform generation
  - quickly update only parts of the circuit


#### API Changes

- `patronus::btor2::parse_file` should take in a `Context` instead of producing one
- `Context` should use `RefCell` to allow expressions to be built with a immutable reference