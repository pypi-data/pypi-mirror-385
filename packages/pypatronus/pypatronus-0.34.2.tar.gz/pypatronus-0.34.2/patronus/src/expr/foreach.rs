// Copyright 2023 The Regents of the University of California
// released under BSD 3-Clause License
// author: Kevin Laeufer <laeufer@berkeley.edu>

use super::{Expr, ExprRef};

pub trait ForEachChild<T: Clone> {
    fn for_each_child(&self, visitor: impl FnMut(&T));
    fn collect_children(&self, children: &mut Vec<T>) {
        self.for_each_child(|c: &T| {
            children.push(c.clone());
        });
    }

    fn num_children(&self) -> usize;
}

impl ForEachChild<ExprRef> for Expr {
    fn for_each_child(&self, mut visitor: impl FnMut(&ExprRef)) {
        match self {
            Expr::BVSymbol { .. } => {}  // no children
            Expr::BVLiteral { .. } => {} // no children
            Expr::BVZeroExt { e, .. } => {
                (visitor)(e);
            }
            Expr::BVSignExt { e, .. } => {
                (visitor)(e);
            }
            Expr::BVSlice { e, .. } => {
                (visitor)(e);
            }
            Expr::BVNot(e, _) => {
                (visitor)(e);
            }
            Expr::BVNegate(e, _) => {
                (visitor)(e);
            }
            Expr::BVEqual(a, b) => {
                (visitor)(a);
                (visitor)(b);
            }
            Expr::BVImplies(a, b) => {
                (visitor)(a);
                (visitor)(b);
            }
            Expr::BVGreater(a, b) => {
                (visitor)(a);
                (visitor)(b);
            }
            Expr::BVGreaterSigned(a, b, _) => {
                (visitor)(a);
                (visitor)(b);
            }
            Expr::BVGreaterEqual(a, b) => {
                (visitor)(a);
                (visitor)(b);
            }
            Expr::BVGreaterEqualSigned(a, b, _) => {
                (visitor)(a);
                (visitor)(b);
            }
            Expr::BVConcat(a, b, _) => {
                (visitor)(a);
                (visitor)(b);
            }
            Expr::BVAnd(a, b, _) => {
                (visitor)(a);
                (visitor)(b);
            }
            Expr::BVOr(a, b, _) => {
                (visitor)(a);
                (visitor)(b);
            }
            Expr::BVXor(a, b, _) => {
                (visitor)(a);
                (visitor)(b);
            }
            Expr::BVShiftLeft(a, b, _) => {
                (visitor)(a);
                (visitor)(b);
            }
            Expr::BVArithmeticShiftRight(a, b, _) => {
                (visitor)(a);
                (visitor)(b);
            }
            Expr::BVShiftRight(a, b, _) => {
                (visitor)(a);
                (visitor)(b);
            }
            Expr::BVAdd(a, b, _) => {
                (visitor)(a);
                (visitor)(b);
            }
            Expr::BVMul(a, b, _) => {
                (visitor)(a);
                (visitor)(b);
            }
            Expr::BVSignedDiv(a, b, _) => {
                (visitor)(a);
                (visitor)(b);
            }
            Expr::BVUnsignedDiv(a, b, _) => {
                (visitor)(a);
                (visitor)(b);
            }
            Expr::BVSignedMod(a, b, _) => {
                (visitor)(a);
                (visitor)(b);
            }
            Expr::BVSignedRem(a, b, _) => {
                (visitor)(a);
                (visitor)(b);
            }
            Expr::BVUnsignedRem(a, b, _) => {
                (visitor)(a);
                (visitor)(b);
            }
            Expr::BVSub(a, b, _) => {
                (visitor)(a);
                (visitor)(b);
            }
            Expr::BVArrayRead { array, index, .. } => {
                (visitor)(array);
                (visitor)(index);
            }
            Expr::BVIte { cond, tru, fals } => {
                (visitor)(cond);
                (visitor)(tru);
                (visitor)(fals);
            }
            Expr::ArraySymbol { .. } => {} // no children
            Expr::ArrayConstant { e, .. } => {
                (visitor)(e);
            }
            Expr::ArrayEqual(a, b) => {
                (visitor)(a);
                (visitor)(b);
            }
            Expr::ArrayStore { array, index, data } => {
                (visitor)(array);
                (visitor)(index);
                (visitor)(data)
            }
            Expr::ArrayIte { cond, tru, fals } => {
                (visitor)(cond);
                (visitor)(tru);
                (visitor)(fals);
            }
        }
    }

    fn num_children(&self) -> usize {
        match self {
            Expr::BVSymbol { .. } => 0,
            Expr::BVLiteral(_) => 0,
            Expr::BVZeroExt { .. } => 1,
            Expr::BVSignExt { .. } => 1,
            Expr::BVSlice { .. } => 1,
            Expr::BVNot(_, _) => 1,
            Expr::BVNegate(_, _) => 1,
            Expr::BVEqual(_, _) => 2,
            Expr::BVImplies(_, _) => 2,
            Expr::BVGreater(_, _) => 2,
            Expr::BVGreaterSigned(_, _, _) => 2,
            Expr::BVGreaterEqual(_, _) => 2,
            Expr::BVGreaterEqualSigned(_, _, _) => 2,
            Expr::BVConcat(_, _, _) => 2,
            Expr::BVAnd(_, _, _) => 2,
            Expr::BVOr(_, _, _) => 2,
            Expr::BVXor(_, _, _) => 2,
            Expr::BVShiftLeft(_, _, _) => 2,
            Expr::BVArithmeticShiftRight(_, _, _) => 2,
            Expr::BVShiftRight(_, _, _) => 2,
            Expr::BVAdd(_, _, _) => 2,
            Expr::BVMul(_, _, _) => 2,
            Expr::BVSignedDiv(_, _, _) => 2,
            Expr::BVUnsignedDiv(_, _, _) => 2,
            Expr::BVSignedMod(_, _, _) => 2,
            Expr::BVSignedRem(_, _, _) => 2,
            Expr::BVUnsignedRem(_, _, _) => 2,
            Expr::BVSub(_, _, _) => 2,
            Expr::BVArrayRead { .. } => 2,
            Expr::BVIte { .. } => 3,
            Expr::ArraySymbol { .. } => 0,
            Expr::ArrayConstant { .. } => 1,
            Expr::ArrayEqual(_, _) => 2,
            Expr::ArrayStore { .. } => 3,
            Expr::ArrayIte { .. } => 3,
        }
    }
}
