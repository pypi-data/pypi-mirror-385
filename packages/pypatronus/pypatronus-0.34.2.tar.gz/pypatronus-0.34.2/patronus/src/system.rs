// Copyright 2023-2024 The Regents of the University of California
// Copyright 2024 Cornell University
// released under BSD 3-Clause License
// author: Kevin Laeufer <laeufer@cornell.edu>

pub mod analysis;
mod serialize;
pub mod transform;
mod transition_system;

pub use transition_system::*;
