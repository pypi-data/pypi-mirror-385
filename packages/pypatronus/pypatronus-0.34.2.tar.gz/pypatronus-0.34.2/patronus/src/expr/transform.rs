// Copyright 2023 The Regents of the University of California
// released under BSD 3-Clause License
// author: Kevin Laeufer <laeufer@berkeley.edu>

use crate::expr::meta::get_fixed_point;
use crate::expr::*;

#[derive(Debug, Copy, Clone, Eq, PartialEq)]
pub enum ExprTransformMode {
    SingleStep,
    FixedPoint,
}

/// transform an expression with a single step (no fixed point) and no persistent cache
#[inline]
pub fn simple_transform_expr(
    ctx: &mut Context,
    e: ExprRef,
    tran: impl FnMut(&mut Context, ExprRef, &[ExprRef]) -> Option<ExprRef>,
) -> ExprRef {
    let mut cache = SparseExprMap::default();
    do_transform_expr(
        ctx,
        ExprTransformMode::SingleStep,
        &mut cache,
        vec![e],
        tran,
    );
    cache[e].unwrap()
}

#[inline]
pub(crate) fn do_transform_expr<T: ExprMap<Option<ExprRef>>>(
    ctx: &mut Context,
    mode: ExprTransformMode,
    transformed: &mut T,
    mut todo: Vec<ExprRef>,
    mut tran: impl FnMut(&mut Context, ExprRef, &[ExprRef]) -> Option<ExprRef>,
) {
    let mut children = Vec::with_capacity(4);

    while let Some(expr_ref) = todo.pop() {
        // check to see if we translated all the children
        children.clear();
        let mut children_changed = false; // track whether any of the children changed
        let mut all_transformed = true; // tracks whether all children have been transformed or if there is more work to do
        ctx[expr_ref].for_each_child(|c| {
            let transformed_child = if mode == ExprTransformMode::FixedPoint {
                get_fixed_point(transformed, *c)
            } else {
                transformed[*c]
            };
            match transformed_child {
                Some(new_child_expr) => {
                    if new_child_expr != *c {
                        children_changed = true; // child changed
                    }
                    children.push(new_child_expr);
                }
                None => {
                    if all_transformed {
                        todo.push(expr_ref);
                    }
                    all_transformed = false;
                    todo.push(*c);
                }
            }
        });
        if !all_transformed {
            continue;
        }

        // call out to the transform
        let tran_res = tran(ctx, expr_ref, &children);
        let new_expr_ref = match tran_res {
            Some(e) => e,
            None => {
                if children_changed {
                    update_expr_children(ctx, expr_ref, &children)
                } else {
                    // if no children changed and the transform does not want to do changes,
                    // we can just keep the old expression
                    expr_ref
                }
            }
        };
        // remember the transformed version
        transformed[expr_ref] = Some(new_expr_ref);

        // in fixed point mode, we might not be done yet
        let is_at_fixed_point = expr_ref == new_expr_ref;
        if mode == ExprTransformMode::FixedPoint
            && !is_at_fixed_point
            && transformed[new_expr_ref].is_none()
        {
            // see if we can further simplify the new expression
            todo.push(new_expr_ref);
        }
    }
}

fn update_expr_children(ctx: &mut Context, expr_ref: ExprRef, children: &[ExprRef]) -> ExprRef {
    let new_expr = match (&ctx[expr_ref], children) {
        (Expr::BVSymbol { .. }, _) => panic!("No children, should never get here."),
        (Expr::BVLiteral { .. }, _) => panic!("No children, should never get here."),
        (Expr::BVZeroExt { by, width, .. }, [e]) => Expr::BVZeroExt {
            e: *e,
            by: *by,
            width: *width,
        },
        (Expr::BVSignExt { by, width, .. }, [e]) => Expr::BVSignExt {
            e: *e,
            by: *by,
            width: *width,
        },
        (Expr::BVSlice { hi, lo, .. }, [e]) => Expr::BVSlice {
            e: *e,
            hi: *hi,
            lo: *lo,
        },
        (Expr::BVNot(_, width), [e]) => Expr::BVNot(*e, *width),
        (Expr::BVNegate(_, width), [e]) => Expr::BVNegate(*e, *width),
        (Expr::BVEqual(_, _), [a, b]) => Expr::BVEqual(*a, *b),
        (Expr::BVImplies(_, _), [a, b]) => Expr::BVImplies(*a, *b),
        (Expr::BVGreater(_, _), [a, b]) => Expr::BVGreater(*a, *b),
        (Expr::BVGreaterSigned(_, _, w), [a, b]) => Expr::BVGreaterSigned(*a, *b, *w),
        (Expr::BVGreaterEqual(_, _), [a, b]) => Expr::BVGreaterEqual(*a, *b),
        (Expr::BVGreaterEqualSigned(_, _, w), [a, b]) => Expr::BVGreaterEqualSigned(*a, *b, *w),
        (Expr::BVConcat(_, _, w), [a, b]) => Expr::BVConcat(*a, *b, *w),
        (Expr::BVAnd(_, _, w), [a, b]) => Expr::BVAnd(*a, *b, *w),
        (Expr::BVOr(_, _, w), [a, b]) => Expr::BVOr(*a, *b, *w),
        (Expr::BVXor(_, _, w), [a, b]) => Expr::BVXor(*a, *b, *w),
        (Expr::BVShiftLeft(_, _, w), [a, b]) => Expr::BVShiftLeft(*a, *b, *w),
        (Expr::BVArithmeticShiftRight(_, _, w), [a, b]) => Expr::BVArithmeticShiftRight(*a, *b, *w),
        (Expr::BVShiftRight(_, _, w), [a, b]) => Expr::BVShiftRight(*a, *b, *w),
        (Expr::BVAdd(_, _, w), [a, b]) => Expr::BVAdd(*a, *b, *w),
        (Expr::BVMul(_, _, w), [a, b]) => Expr::BVMul(*a, *b, *w),
        (Expr::BVSignedDiv(_, _, w), [a, b]) => Expr::BVSignedDiv(*a, *b, *w),
        (Expr::BVUnsignedDiv(_, _, w), [a, b]) => Expr::BVUnsignedDiv(*a, *b, *w),
        (Expr::BVSignedMod(_, _, w), [a, b]) => Expr::BVSignedMod(*a, *b, *w),
        (Expr::BVSignedRem(_, _, w), [a, b]) => Expr::BVSignedRem(*a, *b, *w),
        (Expr::BVUnsignedRem(_, _, w), [a, b]) => Expr::BVUnsignedRem(*a, *b, *w),
        (Expr::BVSub(_, _, w), [a, b]) => Expr::BVSub(*a, *b, *w),
        (Expr::BVArrayRead { width, .. }, [array, index]) => Expr::BVArrayRead {
            array: *array,
            index: *index,
            width: *width,
        },
        (Expr::BVIte { .. }, [cond, tru, fals]) => Expr::BVIte {
            cond: *cond,
            tru: *tru,
            fals: *fals,
        },
        (Expr::ArraySymbol { .. }, _) => panic!("No children, should never get here."),
        (
            Expr::ArrayConstant {
                index_width,
                data_width,
                ..
            },
            [e],
        ) => Expr::ArrayConstant {
            e: *e,
            index_width: *index_width,
            data_width: *data_width,
        },
        (Expr::ArrayEqual(_, _), [a, b]) => Expr::ArrayEqual(*a, *b),
        (Expr::ArrayStore { .. }, [array, index, data]) => Expr::ArrayStore {
            array: *array,
            index: *index,
            data: *data,
        },
        (Expr::ArrayIte { .. }, [cond, tru, fals]) => Expr::ArrayIte {
            cond: *cond,
            tru: *tru,
            fals: *fals,
        },
        (other, _) => {
            todo!("implement code to re-create expression `{other:?}` with updated children")
        }
    };
    ctx.add_expr(new_expr)
}
