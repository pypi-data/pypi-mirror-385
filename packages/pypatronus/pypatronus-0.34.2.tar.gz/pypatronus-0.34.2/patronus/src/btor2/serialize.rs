// Copyright 2023-2024 The Regents of the University of California
// Copyright 2024 Cornell University
// released under BSD 3-Clause License
// author: Kevin Laeufer <laeufer@cornell.edu>
use crate::expr::*;
use crate::system::TransitionSystem;
use std::io::Write;

pub fn serialize(
    ctx: &Context,
    writer: &mut impl Write,
    sys: &TransitionSystem,
) -> std::io::Result<()> {
    Serializer::new(ctx, writer).serialize_sys(sys)
}

const VERSION: Option<&str> = option_env!("CARGO_PKG_VERSION");

pub fn serialize_to_str(ctx: &Context, sys: &TransitionSystem) -> String {
    let mut buf = Vec::new();
    serialize(ctx, &mut buf, sys).expect("Failed to write to string!");
    String::from_utf8(buf).expect("Failed to read string we wrote!")
}

struct Serializer<'a, W: Write> {
    #[allow(dead_code)] // TODO: implement serializer which will use the ctx!
    ctx: &'a Context,
    writer: &'a mut W,
}

impl<'a, W: Write> Serializer<'a, W> {
    fn new(ctx: &'a Context, writer: &'a mut W) -> Self {
        Serializer { ctx, writer }
    }

    fn serialize_sys(&mut self, sys: &TransitionSystem) -> std::io::Result<()> {
        writeln!(
            self.writer,
            "; btor2 description of `{}` generated by patron {}",
            sys.name,
            VERSION.unwrap_or_default()
        )?;

        writeln!(self.writer, "; TODO: implement the btor2 serialization!")?;

        Ok(())
    }
}
