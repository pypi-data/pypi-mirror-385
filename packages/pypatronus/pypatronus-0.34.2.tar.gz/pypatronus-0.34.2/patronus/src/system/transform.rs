// Copyright 2023-2024 The Regents of the University of California
// Copyright 2024 Cornell University
// released under BSD 3-Clause License
// author: Kevin Laeufer <laeufer@cornell.edu>

use super::TransitionSystem;
use crate::btor2::{DEFAULT_INPUT_PREFIX, DEFAULT_STATE_PREFIX};
use crate::expr::*;
use rustc_hash::FxHashMap;

/** Remove any inputs named `_input_[...]` and replace their use with a literal zero.
* This essentially gets rid of all undefined value modelling by yosys.
 */
pub fn replace_anonymous_inputs_with_zero(ctx: &mut Context, sys: &mut TransitionSystem) {
    // find and remove inputs
    let mut replace_map = FxHashMap::default();
    sys.inputs.retain(|&input| {
        let name = ctx.get_symbol_name(input).unwrap();
        if name.starts_with(DEFAULT_INPUT_PREFIX) || name.starts_with(DEFAULT_STATE_PREFIX) {
            let replacement = match input.get_type(ctx) {
                Type::BV(width) => ctx.zero(width),
                Type::Array(tpe) => ctx.zero_array(tpe),
            };
            replace_map.insert(input, replacement);
            false
        } else {
            true
        }
    });

    // replace any use of the input with zero
    do_transform(
        ctx,
        sys,
        ExprTransformMode::SingleStep,
        |_ctx, expr, _children| replace_map.get(&expr).cloned(),
    );
}

/// Applies simplifications to the expressions used in the system.
pub fn simplify_expressions(ctx: &mut Context, sys: &mut TransitionSystem) {
    do_transform(ctx, sys, ExprTransformMode::FixedPoint, simplify);
}

pub fn do_transform(
    ctx: &mut Context,
    sys: &mut TransitionSystem,
    mode: ExprTransformMode,
    tran: impl FnMut(&mut Context, ExprRef, &[ExprRef]) -> Option<ExprRef>,
) {
    // update all expressions used in the transition system
    let todo = sys.get_all_exprs();
    let mut transformed = DenseExprMetaData::default();
    do_transform_expr(ctx, mode, &mut transformed, todo, tran);

    // update transition system signals to point to updated expressions
    sys.update_expressions(|old_expr| {
        if mode == ExprTransformMode::FixedPoint {
            get_fixed_point(&mut transformed, old_expr)
        } else {
            transformed[old_expr]
        }
    });
}
