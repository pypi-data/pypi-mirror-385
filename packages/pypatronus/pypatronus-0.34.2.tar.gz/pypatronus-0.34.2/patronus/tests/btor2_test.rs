// Copyright 2023-2024 The Regents of the University of California
// Copyright 2024 Cornell University
// released under BSD 3-Clause License
// author: Kevin Laeufer <laeufer@cornell.edu>

use patronus::btor2;
use patronus::expr::*;
use patronus::system::transform::{replace_anonymous_inputs_with_zero, simplify_expressions};

const COUNT_2: &str = r#"
1 sort bitvec 3
2 zero 1
3 state 1
4 init 1 3 2
5 one 1
6 add 1 3 5
7 next 1 3 6
8 ones 1
9 sort bitvec 1
10 eq 9 3 8
11 bad 10
"#;

#[test]
fn parse_count2() {
    let mut ctx = Context::default();
    let sys = btor2::parse_str(&mut ctx, COUNT_2, Some("count2")).unwrap();
    insta::assert_snapshot!(sys.serialize_to_str(&ctx));
}

#[test]
fn serialize_count2() {
    let mut ctx = Context::default();
    let sys = btor2::parse_str(&mut ctx, COUNT_2, Some("count2")).unwrap();
    insta::assert_snapshot!(skip_first_line(&btor2::serialize_to_str(&ctx, &sys)));
}

const PASS_THROUGH: &str = r#"
1 sort bitvec 1
2 input 1 i
3 output 2 o
"#;

#[test]
fn serialize_pass_through() {
    let mut ctx = Context::default();
    let sys = btor2::parse_str(&mut ctx, PASS_THROUGH, Some("pass")).unwrap();
    insta::assert_snapshot!(sys.serialize_to_str(&ctx));
}

#[test]
fn parse_quiz1() {
    let (ctx, sys) = btor2::parse_file("../inputs/chiseltest/Quiz1.btor").unwrap();
    insta::assert_snapshot!(sys.serialize_to_str(&ctx));
}

#[test]
fn serialize_quiz1() {
    let (ctx, sys) = btor2::parse_file("../inputs/chiseltest/Quiz1.btor").unwrap();
    insta::assert_snapshot!(skip_first_line(&btor2::serialize_to_str(&ctx, &sys)));
}

#[test]
fn parse_instrumented_decoder() {
    let (ctx, sys) =
        btor2::parse_file("../inputs/repair/decoder_3_to_8.instrumented.btor").unwrap();
    insta::assert_snapshot!(sys.serialize_to_str(&ctx));
}

#[test]
fn parse_sdram() {
    let (mut ctx, mut sys) =
        btor2::parse_file("../inputs/repair/sdram_controller.original.btor").unwrap();
    replace_anonymous_inputs_with_zero(&mut ctx, &mut sys);
    insta::assert_snapshot!(sys.serialize_to_str(&ctx));
}

#[test]
fn parse_sdram_and_remove_anonymous_inputs() {
    let (mut ctx, mut sys) =
        btor2::parse_file("../inputs/repair/sdram_controller.original.btor").unwrap();
    replace_anonymous_inputs_with_zero(&mut ctx, &mut sys);
    insta::assert_snapshot!(sys.serialize_to_str(&ctx));
}

#[test]
fn parse_sdram_and_simplify_expressions() {
    let (mut ctx, mut sys) =
        btor2::parse_file("../inputs/repair/sdram_controller.original.btor").unwrap();
    replace_anonymous_inputs_with_zero(&mut ctx, &mut sys);
    simplify_expressions(&mut ctx, &mut sys);
    // println!("{}", sys.serialize_to_str(&ctx));
    insta::assert_snapshot!(sys.serialize_to_str(&ctx));
}

/// Regression test for a bug that would prevent "_input_0" from being listed as a system input.
/// The reason was that we did not handle inputs that are also outputs well.
#[test]
fn parse_sha3_keccak_and_check_that_all_anonymous_inputs_are_there() {
    let (ctx, sys) =
        btor2::parse_file("../inputs/repair/sha3_keccak.w2.replace_variables.btor").unwrap();
    insta::assert_snapshot!(sys.serialize_to_str(&ctx));
}

/// Regression test to make sure that when an input that is also an output is set to zero,
/// it is maintained.
#[test]
fn parse_sha3_keccak_and_replace_anonymous_inputs() {
    let (mut ctx, mut sys) =
        btor2::parse_file("../inputs/repair/sha3_keccak.w2.replace_variables.btor").unwrap();
    replace_anonymous_inputs_with_zero(&mut ctx, &mut sys);
    insta::assert_snapshot!(sys.serialize_to_str(&ctx));
}

fn skip_first_line(value: &str) -> &str {
    match value.char_indices().find(|(_, c)| *c == '\n') {
        None => "",
        Some((char_ii, _)) => &value[(char_ii + 1)..],
    }
}

#[test]
#[ignore] // TODO: this test was randomly failing on the CI, but passing locally.
fn parse_lakeroad_dsp48_e2() {
    let (mut ctx, mut sys) = btor2::parse_file("../inputs/lakeroad/DSP48E2.btor").unwrap();
    replace_anonymous_inputs_with_zero(&mut ctx, &mut sys);
    simplify_expressions(&mut ctx, &mut sys);
    insta::assert_snapshot!(sys.serialize_to_str(&ctx));
}

const INVERTER: &str = r#"
; BTOR description generated by Yosys 0.51 (git sha1 c4b5190229616f7ebf8197f43990b4429de3e420, clang++ 16.0.0 -fPIC -O3) for module inverter_d0.
1 sort bitvec 32
2 input 1 a ; /Users/nikilshyamsunder/capra/protocols/examples/inverters/inverter_d0.v:2.19-2.20
3 not 1 2 $not$/Users/nikilshyamsunder/capra/protocols/examples/inverters/inverter_d0.v:6$1 ; /Users/nikilshyamsunder/capra/protocols/examples/inverters/inverter_d0.v:6.16-6.18
4 output 3 s ; /Users/nikilshyamsunder/capra/protocols/examples/inverters/inverter_d0.v:3.19-3.20
; end of yosys output
"#;

#[test]
fn serialize_inverter() {
    let mut ctx = Context::default();
    let sys = btor2::parse_str(&mut ctx, INVERTER, Some("inverter_d0")).unwrap();
    // check to make sure the output and input names are correct
    let inputs: Vec<_> = sys
        .inputs
        .iter()
        .map(|i| ctx.get_symbol_name(*i).unwrap().to_string())
        .collect();
    let outputs: Vec<_> = sys
        .outputs
        .iter()
        .map(|o| ctx[o.name].to_string())
        .collect();
    assert_eq!(inputs, ["a"]);
    assert_eq!(outputs, ["s"]);
    insta::assert_snapshot!(sys.serialize_to_str(&ctx));
}

#[test]
fn parse_llsdspi_bug_c1_instrumented() {
    let (mut ctx, mut sys) =
        btor2::parse_file("../inputs/repair/llsdspi_bug_c1.instrumented.btor").unwrap();
    replace_anonymous_inputs_with_zero(&mut ctx, &mut sys);
    simplify_expressions(&mut ctx, &mut sys);
    println!("{}", sys.serialize_to_str(&ctx));
}
