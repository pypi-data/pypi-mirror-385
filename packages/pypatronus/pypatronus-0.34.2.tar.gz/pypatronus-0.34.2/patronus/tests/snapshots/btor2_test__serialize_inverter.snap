---
source: patronus/tests/btor2_test.rs
expression: sys.serialize_to_str(&ctx)
---
inverter_d0
input a : bv<32>
output s : bv<32> = not(a)

